pkg/ruler/ruler.go

```diff
@@ -1164,6 +1285,8 @@ func (r *Ruler) getShardedRules(ctx context.Context, userID string, rulesRequest
 		}
 	}
 	// Concurrently fetch rules from all rulers.
+	ctx, cancel := context.WithTimeout(ctx, r.cfg.ListRulesFanoutTimeout)
```

Comment:
This will be a behavior change for all Cortex users. Can we just use the API request's own timeout?

Reply:
If the client sets the timeout, then this timeout/deadline will also be the same and it will fail the overall API request for the client. The desired behavior is not to fail the whole API request but instead be fault tolerant if a subset of Rulers do not respond in a certain amount of time. Ideally,  `r.cfg.ListRulesFanoutTimeout` should be less than the API request's timeout. If we do not want to change existing behavior, we can either set the default timeout to be a much bigger number (max int) or wrap this around the `enable_ha_evaluation` flag or perhaps there is a better option. Open to recommendations

Reply:
+1 to bens comment... I think the solution here is not to set a timeout.. instead, we should return as soon as we have a complete answer (quorum?)

Reply:
I'm curious, is this line change related to minimizing missed evaluations or is it unrelated? If unrelated and you are solving another problem I would rather have this in a different pr, just because this can introduce regressions.

Reply:
> +1 to bens comment... I think the solution here is not to set a timeout.. instead, we should return as soon as we have a complete answer (quorum?)

When Ruler API HA is enabled and the rules are backed up, the backed up rules do not contain any state. In other words, it does not contain any alerts. To get a complete answer means, we have to get Rules from 100% of the Rulers. 

As an example, 3 AZs, RF = 3, ruler count per AZ = 3 (total = 9). Assume 1 ruler in AZ 2 is unhealthy. What we want is results from 8 rulers. Similarly if we have 2 bad rulers, then we need results from 7. In this scenario, we we can tolerate 2 AZs being down (RF=3) but if we bail as soon as results from 1 AZ are retrieved, then the results will contain a lot of backup rules which does not contain alerts

I moved the timeout inside the `concurrency.ForEach` so that each `Rules()` fan-out call will get its own timeout rather than the timeout being set for the whole `getShardedRules`

Let me know if this is acceptable or if there are other ideas to handle this scenario